include/dice/template-library/pool_allocator.hpp

```diff
@@ -0,0 +1,205 @@
+#ifndef DICE_TEMPLATELIBRARY_POOLALLOCATOR_HPP
+#define DICE_TEMPLATELIBRARY_POOLALLOCATOR_HPP
+
+#include <boost/pool/pool.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <type_traits>
+
+namespace dice::template_library {
+
+	/**
+	 * A memory pool or arena that is efficient for allocations which are smaller or equal in size
+	 * for one of `bucket_sizes...`.
+	 *
+	 * The implementation consists of one arena per provided bucket size.
+	 * An allocation will be placed into the first bucket where it can fit.
+	 * Allocations that do not fit into any bucket are fulfilled with calls to `new`.
```

Comment:
What's the reasoning behind allowing allocations that do not fit into any of the buckets?
Does this not beat the purpose of having the pools?

Reply:
The rationale behind this is that you may have fairly large allocations that happen infrequently enough that it does not make sense to have a pool for those.

For our specific use case it would be for the sparse maps (Reminder: sparse maps are basically `vector<array<T, 64>>`). For the outer vector you cannot really create a pool because you don't know how large the allocation may grow.

We need this allocator to be able to do this as well because we can only pass in a single allocator.